core/blockchain.go

```diff
@@ -1615,19 +1616,49 @@ func (bc *BlockChain) writeBlockWithState(block *types.Block, receipts []*types.
 	// If the total difficulty is higher than our known, add it to the canonical chain
 	// Second clause in the if statement reduces the vulnerability to selfish mining.
 	// Please refer to http://www.cs.cornell.edu/~ie53/publications/btcProcFC.pdf
-	reorg := externTd.Cmp(localTd) > 0
+	reorg := externTd.Cmp(localTd) > 0 // EIP-3436 (1): Choose the block with the most total difficulty.
 	currentBlock = bc.CurrentBlock()
 	if !reorg && externTd.Cmp(localTd) == 0 {
 		// Split same-difficulty blocks by number, then preferentially select
 		// the block generated by the local miner as the canonical block.
 		if block.NumberU64() < currentBlock.NumberU64() {
-			reorg = true
-		} else if block.NumberU64() == currentBlock.NumberU64() {
+			reorg = true // EIP-3436 (2): Choose the block with the lowest block number.
+
+		} else if block.NumberU64() == currentBlock.NumberU64() && bc.Config().GetConsensusEngineType().IsEthash() {
```

Comment:
Add a restriction to this scope limiting it to `Ethash`.

When this scope is applied to `Clique` (as originally), the always-false return value of `bc.shouldPreserve` yielded a truthy `reorg` value indescriminately 50% of the time.